### PR TITLE
introduce the variable of allow_append_initplan_for_function_scan

### PR DESCRIPTION
**Pull request by SmartKeyerror**
This PR is trying to fix the issue of #16679.

Since #9542 introduced a new location option for function: EXECUTE ON INITPLAN 
and did some patches at #10156, there are still some bugs to address.

Theoretically, if a SubLink could transform into InitPlan, then we should not consider 
invoking the function `append_initplan_for_function_scan()` to make another initplan, 
this is the RCA of that issue. Just like the below plan:

```
create language plpython3u;

create or replace function hello_initplan() returns setof text as $$
return ["hello"]
$$ language plpython3u
execute on initplan;

create table u (a int, b int);

explain (analyze,costs off) select * from u where a::text = (select f from hello_initplan() as f);
                                         QUERY PLAN                                          
---------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3) (actual time=0.325..0.325 rows=0 loops=1)
   InitPlan 2 (returns $1)  (slice2)
     ->  Function Scan on hello_initplan f_1 (actual time=0.014..0.015 rows=1 loops=1)
           InitPlan 1 (returns $0)  (slice3)
             ->  Function Scan on hello_initplan f (actual time=0.049..0.049 rows=1 loops=1)
   ->  Seq Scan on u (actual time=0.000..0.016 rows=0 loops=1)
         Filter: ((a)::text = $1)
```

You can see there are 2 InitPlans that executed the same thing.

so in this PR I introduced a variable of `allow_append_initplan_for_function_scan` to control
it's behavior, although we don't like global variable, but I have not other better solution.

So if you have more graceful idea, please let me know, even just a thought, we could discuss
it, thanks a lot.

### CODE REVIEW

File: src/backend/cdb/cdbpath.c [hunk 1]
```diff
@@ -2669,3 +2681,21 @@ can_elide_explicit_motion(PlannerInfo *root, Index rti, Path *subpath,
...
+void
+set_not_allow_append_initplan_for_function_scan()
+{
+	allow_append_initplan_for_function_scan = false;
+}
+
+void
+set_allow_append_initplan_for_function_scan()
+{
+	allow_append_initplan_for_function_scan = true;
```
[the remark this answers is not in the document]
Comment (SmartKeyerror) [reply src/backend/cdb/cdbpath.c:2694]: Nope, it's a hard code

File: src/backend/optimizer/plan/subselect.c [hunk 2]
```diff
@@ -3513,3 +3521,22 @@ SS_make_initplan_from_plan(PlannerInfo *root,
...
+
+
+bool
+splan_is_initplan(List *plan_params, SubLinkType subLinkType)
+{
+	/*
+	 * un-correlated or undirect correlated plans of EXISTS, EXPR, ARRAY,
+	 * ROWCOMPARE, or MULTIEXPR types can be used as initPlans.
+	 */
+	if (plan_params == NIL && (
```
[the remark this answers is not in the document]
Comment (SmartKeyerror) [reply src/backend/optimizer/plan/subselect.c:3533]: Yeah, but this is another new function, we need to complete all the factors.

**Comment by charliettxx** [comment]
Is that possible to add the flag `allow_append_initplan_for_function_scan` to PlannerConfig which also has other flags to be set in `make_subplan`? so we could avoid adding global variables in planner. If we cannot avoid the global variable, I suggest we could add it to `PlannerGlobal`, which is the global variable for planner.

**Comment by SmartKeyerror** [comment]
> Is that possible to add the flag allow_append_initplan_for_function_scan to PlannerConfig which also has other flags to be set in make_subplan? 

I'm not sure this is a better idea, what's your opinion? @adam8157

**Comment by SmartKeyerror** [comment]
dev pipeline: https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/lzl-narrow-initplan-function?group=all

File: src/backend/cdb/cdbpath.c [hunk 3]
```diff
@@ -2669,3 +2681,21 @@ can_elide_explicit_motion(PlannerInfo *root, Index rti, Path *subpath,
 
 	return false;
 }
+
+void
+set_not_allow_append_initplan_for_function_scan()
```
Comment (adam8157) [inline src/backend/cdb/cdbpath.c:2686]: `unset_...`

**Review by adam8157 — APPROVED** [review]
I prefer the current implementation, the flag is a temporary switch, but the `PlannerConfig` or `PlannerGlobal` is for the whole planning stage.

File: src/backend/optimizer/plan/subselect.c [hunk 4]
```diff
@@ -428,10 +431,15 @@ make_subplan(PlannerInfo *root, Query *orig_subquery,
...
 	subroot->curSlice->gangType = GANGTYPE_UNALLOCATED;
 
+	if (splan_is_initplan(plan_params, subLinkType))
+		set_not_allow_append_initplan_for_function_scan();
+
 	plan = create_plan(subroot, best_path, subroot->curSlice);
 	/* Decorate the top node of the plan with a Flow node. */
 	plan->flow = cdbpathtoplan_create_flow(subroot, best_path->locus);
 
+	set_allow_append_initplan_for_function_scan();
```
Comment (avamingli) [inline src/backend/optimizer/plan/subselect.c:441]: Hi, how about adding an assertion here,  to make sure that variable has been recovered after hacked .

**Pull request merged**